File: python/test/test_impedance_spectroscopy.py
# Copyright (c) 2016, the Cap authors.
#
# This file is subject to the Modified BSD License and may not be distributed
# without copyright and license information. Please refer to the file LICENSE
# for the text and further information on this license. 

from pycap import PropertyTree, EnergyStorageDevice, Experiment,\
                  ECLabAsciiFile
from pycap import measure_impedance_spectrum, retrieve_impedance_spectrum,\
                  fourier_analysis, initialize_data
from numpy import inf, linalg, real, imag, pi, log10, absolute, angle
from numpy import array, testing
from warnings import catch_warnings, simplefilter
from h5py import File
import unittest

R = 50.0e-3   # ohm
R_L = 500.0   # ohm
C = 3.0       # farad

ptree = PropertyTree()
ptree.put_string('type', 'ElectrochemicalImpedanceSpectroscopy')
ptree.put_double('frequency_upper_limit', 1e+4)
ptree.put_double('frequency_lower_limit', 1e-6)
ptree.put_int('steps_per_decade', 3)
ptree.put_int('steps_per_cycle', 1024)
ptree.put_int('cycles', 2)
ptree.put_int('ignore_cycles', 1)
ptree.put_double('dc_voltage', 0)
ptree.put_string('harmonics', '3')
ptree.put_string('amplitudes', '5e-3')
ptree.put_string('phases', '0')
eis = Experiment(ptree)

def setup_expertiment():
    eis_database = PropertyTree()
    eis_database.put_double('frequency_upper_limit', 1e+4)
    eis_database.put_double('frequency_lower_limit', 1e-6)
    eis_database.put_int('steps_per_decade', 3)
    eis_database.put_int('steps_per_cycle', 1024)
    eis_database.put_int('cycles', 2)
    eis_database.put_int('ignore_cycles', 1)
    eis_database.put_double('dc_voltage', 0)
    eis_database.put_string('harmonics', ' 3')
    eis_database.put_string('amplitudes', ' 5e-3')
    eis_database.put_string('phases', '0')

    return eis_database


class capImpedanceSpectroscopyTestCase(unittest.TestCase):
    def test_fourier_analysis(self):
        ptree = PropertyTree()
        ptree.put_int('steps_per_cycle', 3)
        ptree.put_int('cycles', 1)
        ptree.put_int('ignore_cycles', 0)
        ptree.put_string('harmonics', '1')
        # uninitialized data
        data = {}
        self.assertRaises(KeyError, fourier_analysis, data, ptree)
        # empty data
        data = initialize_data()
        self.assertRaises(IndexError, fourier_analysis, data, ptree)
        # bad data
        data['time'] = array([1, 2, 3], dtype=float)
        data['current'] = array([4, 5, 6], dtype=float)
        data['voltage'] = array([7, 8], dtype=float)
        self.assertRaises(AssertionError, fourier_analysis, data, ptree)
        # poor data (size not a power of 2)
        data['voltage'] = array([7, 8, 9], dtype=float)
        with catch_warnings():
            simplefilter("error")
            self.assertRaises(RuntimeWarning, fourier_analysis, data, ptree)
        # data unchanged after analyze
        dummy = array([1, 2, 3, 4, 5, 6, 7, 8], dtype=float)
        data['time'] = dummy
        data['current'] = dummy
        data['voltage'] = dummy
        # ptree needs to be updated
        self.assertRaises(AssertionError, fourier_analysis, data, ptree)
        ptree.put_int('steps_per_cycle', 4)
        ptree.put_int('cycles', 2)
        ptree.put_int('ignore_cycles', 0)
        fourier_analysis(data, ptree)
        try:
            testing.assert_array_equal(data['time'], dummy)
            testing.assert_array_equal(data['current'], dummy)
            testing.assert_array_equal(data['voltage'], dummy)
        except AssertionError:
            self.fail('data should not be changed by the fourier analyzis')

    def test_retrieve_data(self):
        # TODO: no need to use a device here
        device_database = PropertyTree()
        device_database.put_string('type', 'SeriesRC')
        device_database.put_double('series_resistance', R)
        device_database.put_double('capacitance', C)
        device = EnergyStorageDevice(device_database)
        eis_database  = setup_expertiment()
        eis_database.put_int('steps_per_decade', 1)
        eis_database.put_int('steps_per_cycle', 64)
        eis_database.put_int('cycles', 2)
        eis_database.put_int('ignore_cycles', 1)
        fout = File('trash.hdf5', 'w')
        spectrum_data = measure_impedance_spectrum(device, eis_database, fout)
        fout.close()
        fin = File('trash.hdf5', 'r')
        retrieved_data = retrieve_impedance_spectrum(fin)
        fin.close()
        print(spectrum_data['impedance']-retrieved_data['impedance'])
        print(retrieved_data)
        self.assertEqual(linalg.norm(spectrum_data['frequency'] -
                                     retrieved_data['frequency'], inf), 0.0)
        # not sure why we don't get equality for the impedance
        self.assertLess(linalg.norm(spectrum_data['impedance'] -
                                    retrieved_data['impedance'], inf), 1e-10)

    def test_verification_with_equivalent_circuit(self):
        # setup equivalent circuit database
        device_database = PropertyTree()
        device_database.put_double('series_resistance', R)
        device_database.put_double('parallel_resistance', R_L)
        device_database.put_double('capacitance', C)
        # analytical solutions
        Z = {}
        Z['SeriesRC'] = lambda f : R+1/(1j*C*2*pi*f)
        Z['ParallelRC'] = lambda f : R+R_L/(1+1j*R_L*C*2*pi*f)
        for device_type in ['SeriesRC', 'ParallelRC']:
            # create a device
            device_database.put_string('type', device_type)
            device = EnergyStorageDevice(device_database)
            # setup experiment and measure
            eis.reset()
            eis.run(device)
            f = eis._data['frequency']
            Z_computed = eis._data['impedance']
            # compute the exact solution
            Z_exact = Z[device_type](f)
            # ensure the error is small
            max_phase_error_in_degree = linalg.norm(
                angle(Z_computed)*180/pi - angle(Z_exact)*180/pi,
                inf)
            max_magniture_error_in_decibel = linalg.norm(
                20*log10(absolute(Z_exact)) - 20*log10(absolute(Z_computed)),
                inf)
            print(device_type)
            print('-- max_phase_error_in_degree = {0}'.format(max_phase_error_in_degree))
            print('-- max_magniture_error_in_decibel = {0}'.format(max_magniture_error_in_decibel))
            self.assertLessEqual(max_phase_error_in_degree, 1)
            self.assertLessEqual(max_magniture_error_in_decibel, 0.2)

    def test_export_eclab_ascii_format(self):
            pass

if __name__ == '__main__':
    unittest.main()

File: cpp/test/test_impedance_spectroscopy.cc
/* Copyright (c) 2016, the Cap authors.
 *
 * This file is subject to the Modified BSD License and may not be distributed
 * without copyright and license information. Please refer to the file LICENSE
 * for the text and further information on this license.
 */

#define BOOST_TEST_MODULE TestImpedanceSpectroscopy
#define BOOST_TEST_MAIN
#include <cap/energy_storage_device.h>
#include <cap/electrochemical_impedance_spectroscopy.h>
#include <cap/utils.h>
#include <boost/format.hpp>
#include <boost/foreach.hpp>
#include <boost/property_tree/ptree.hpp>
#include <boost/property_tree/info_parser.hpp>
#include <boost/math/constants/constants.hpp>
#include <boost/test/unit_test.hpp>
#include <gsl/gsl_fft_real.h>
#include <iostream>
#include <fstream>
#include <complex>
#include <iterator>
#include <algorithm>

namespace cap
{

std::function<std::vector<std::tuple<double, std::complex<double>>>(
    std::shared_ptr<boost::property_tree::ptree const>)>
get_compute_exact(
    std::shared_ptr<boost::property_tree::ptree const> dev_database,
    std::shared_ptr<boost::property_tree::ptree const> eis_database)
{
  std::string const device_type = dev_database->get<std::string>("type");
  double const series_resistance =
      dev_database->get<double>("series_resistance");
  double const parallel_resistance =
      dev_database->get<double>("parallel_resistance");
  double const capacitance = dev_database->get<double>("capacitance");

  std::vector<int> const harmonics =
      cap::to_vector<int>(eis_database->get<std::string>("harmonics"));
  double const pi = boost::math::constants::pi<double>();

  if (device_type.compare("SeriesRC") == 0)
  {
    return [series_resistance, capacitance, pi, harmonics](
        std::shared_ptr<boost::property_tree::ptree const> database)
    {
      std::vector<std::tuple<double, std::complex<double>>> results;
      double const frequency = database->get<double>("frequency");
      for (int k : harmonics)
        results.emplace_back(std::make_tuple(
            k * frequency,
            series_resistance +
                1.0 / std::complex<double>(0.0, capacitance * 2.0 * pi * k *
                                                    frequency)));
      return results;
    };
  }
  else if (device_type.compare("ParallelRC") == 0)
  {
    return [series_resistance, parallel_resistance, capacitance, pi, harmonics](
        std::shared_ptr<boost::property_tree::ptree const> database)
    {
      std::vector<std::tuple<double, std::complex<double>>> results;
      double const frequency = database->get<double>("frequency");
      for (int k : harmonics)
        results.emplace_back(std::make_tuple(
            k * frequency, series_resistance +
                               parallel_resistance /
                                   std::complex<double>(
                                       1.0, parallel_resistance * capacitance *
                                                2.0 * pi * k * frequency)));
      return results;
    };
  }
  else
  {
    throw std::runtime_error("invalid device type " + device_type);
  }
}

void scan(std::shared_ptr<cap::EnergyStorageDevice> dev,
          std::shared_ptr<boost::property_tree::ptree const> dev_database,
          std::shared_ptr<boost::property_tree::ptree const> eis_database,
          std::ostream &os = std::cout)
{
  double const frequency_upper_limit =
      eis_database->get<double>("frequency_upper_limit");
  double const frequency_lower_limit =
      eis_database->get<double>("frequency_lower_limit");
  int const steps_per_decade = eis_database->get<int>("steps_per_decade");
  double const pi = boost::math::constants::pi<double>();
  std::vector<int> const harmonics =
      cap::to_vector<int>(eis_database->get<std::string>("harmonics"));
  double const percent_tolerance =
      eis_database->get<double>("percent_tolerance");

  auto compute_exact = get_compute_exact(dev_database, eis_database);

  double expe_frequency;
  std::complex<double> expe_impedance;
  double theo_frequency;
  std::complex<double> theo_impedance;

  std::shared_ptr<boost::property_tree::ptree> tmp =
      std::make_shared<boost::property_tree::ptree>(*eis_database);

  os << "# impedance Z(f) = R + i X \n";
  os << boost::format("# %22s  %22s  %22s  %22s  %22s  \n") %
            "frequency_f_[Hz]" % "resistance_R_[ohm]" % "reactance_X_[ohm]" %
            "magnitude_|Z|_[ohm]" % "phase_arg(Z)_[degree]";

  for (double frequency = frequency_upper_limit;
       frequency >= frequency_lower_limit;
       frequency /= std::pow(10.0, 1.0 / steps_per_decade))
  {
    tmp->put("frequency", frequency);
    auto expe_results = measure_impedance(dev, tmp);
    auto theo_results = compute_exact(tmp);
    for (std::size_t k = 0; k < harmonics.size(); ++k)
    {
      auto it = expe_results.begin();
      std::advance(it, k);
      std::tie(expe_frequency, expe_impedance) = *it;
      std::tie(theo_frequency, theo_impedance) = theo_results[k];

      BOOST_CHECK_CLOSE(expe_frequency, theo_frequency, percent_tolerance);
      BOOST_CHECK_CLOSE(expe_impedance.real(), theo_impedance.real(),
                        percent_tolerance);
      BOOST_CHECK_CLOSE(expe_impedance.imag(), theo_impedance.imag(),
                        percent_tolerance);
      BOOST_CHECK_CLOSE(std::abs(expe_impedance), std::abs(theo_impedance),
                        percent_tolerance);
      BOOST_CHECK_CLOSE(std::arg(expe_impedance), std::arg(theo_impedance),
                        percent_tolerance);

      os << boost::format("  %22.15e  %22.15e  %22.15e  %22.15e  %22.15e  "
                          "%22.15e  %22.15e  %22.15e  %22.15e  %22.15e \n") %
                expe_frequency % expe_impedance.real() % expe_impedance.imag() %
                std::abs(expe_impedance) %
                (std::arg(expe_impedance) * 180.0 / pi) % theo_frequency %
                theo_impedance.real() % theo_impedance.imag() %
                std::abs(theo_impedance) %
                (std::arg(theo_impedance) * 180.0 / pi);
    }
  }
}

} // end namespace cap

BOOST_AUTO_TEST_CASE(test_impedance_spectroscopy)
{
  // parse input file
  std::shared_ptr<boost::property_tree::ptree> input_database =
      std::make_shared<boost::property_tree::ptree>();
  boost::property_tree::info_parser::read_info(
      "input_impedance_spectroscopy.info", *input_database);

  // build an energy storage system
  std::shared_ptr<boost::property_tree::ptree> device_database =
      std::make_shared<boost::property_tree::ptree>(
          input_database->get_child("device"));
  std::shared_ptr<cap::EnergyStorageDevice> device =
      cap::EnergyStorageDevice::build(boost::mpi::communicator(),
                                      *device_database);

  // measure its impedance
  std::fstream fout;
  fout.open("computed_vs_exact_impedance_spectroscopy_data", std::fstream::out);

  std::shared_ptr<boost::property_tree::ptree> impedance_spectroscopy_database =
      std::make_shared<boost::property_tree::ptree>(
          input_database->get_child("impedance_spectroscopy"));
  cap::scan(device, device_database, impedance_spectroscopy_database, fout);

  fout.close();
}
